[WIP] add readme when .pytest_cache is created

avirlrma opened the pull request:
WIP PR for #3519.

avirlrma commented:
Hi @nicoddemus 
I have made a PR with changes as discussed. I still have to add tests for this and work on the Environment variable that you mentioned in the ticket description

nicoddemus commented on src/_pytest/cacheprovider.py:
```diff
@@ -85,6 +85,13 @@ def set(self, key, value):
                python types, including nested types
                like e. g. lists of dictionaries.
         """
+        content_readme = 'test'
```
I suggest to move this block of code to its own function to improve readability, something like `_ensure_readme()` perhaps?

avirlrma replied:
oh yes! That was so evident. Stupid me. Sorry.

nicoddemus commented on src/_pytest/cacheprovider.py:
```diff
@@ -85,6 +85,13 @@ def set(self, key, value):
+        content_readme = 'test'
+        if os.path.isdir(self._cachedir):
+            readme_path = os.path.join(self._cachedir, 'README.md')
+            if os.path.isfile(readme_path) is False:
```
The idiomatic way is to use `if not os.path.isfile(readme_path)`. Also, it is probably easier to use the [`py.path` API](http://py.readthedocs.io/en/latest/path.html) directly:

```python
        if self._cachedir.check(dir=True):
            readme_path = self._cachedir.join('README.md')
            if not readme_path.check(file=True):
                readme_path.write(content_readme)
```

avirlrma commented:
@nicoddemus 
I made the changes as suggested by you. Sorry for overlooking the `black` part in guidelines.
Also I moved the call to `_ensure_readme()` to the bottom of `set()`,so that the permissions `try and except` block can do their work and the code won't be repeated.

I have updated the above checklist.Please check if it's all that neccesary and enough.

nicoddemus commented on src/_pytest/cacheprovider.py:
```diff
@@ -103,6 +103,14 @@ def set(self, key, value):
             with f:
                 self.trace("cache-write %s: %r" % (key, value))
                 json.dump(value, f, indent=2, sort_keys=True)
+        self._ensure_readme()
```
This probably should be indented with the `with f` statement above; this way we ensure we have written the key/value and will be able to write the readme file as well.

nicoddemus commented:
>Sorry for overlooking the black part in guidelines.

No need to be sorry at all, that's what PRs are for! 👍

nicoddemus commented on src/_pytest/cacheprovider.py:
```diff
@@ -103,6 +103,14 @@ def set(self, key, value):
+        self._ensure_readme()
+
+    def _ensure_readme(self):
+        content_readme = "test"
```
We need to include some actual contents. How about:

```markdown
# pytest cache directory #

This directory contains data from the pytest's cache plugin, which provides the `--lf` and `--ff` options, as well as the `cache` fixture. 

**Do not** commit this to version control.

See [the docs](https://docs.pytest.org/en/latest/cache.html) for more information.
```

avirlrma replied:
looks good.I will add it in

RonnyPfannschmidt commented:
please rebase, those merges are a massive problem

avirlrma commented:
@RonnyPfannschmidt @nicoddemus I couldn't solve what had happened. Anyways I'm made a new PR #3608,sorry for the trouble caused.
